runtime/vm/ContinuationHelpers.cpp

```diff
@@ -0,0 +1,189 @@
+#else
+#define VMTHR_INITIAL_STACK_SIZE vm->stackSize
+#endif
+
+	if ((stack = allocateJavaStack(vm, VMTHR_INITIAL_STACK_SIZE, NULL)) == NULL) {
+		vm->internalVMFunctions->setNativeOutOfMemoryError(currentThread, 0, 0);
+		result = FALSE;
+		goto end;
+	}
+
+#undef VMTHR_INITIAL_STACK_SIZE
+
+	continuation->stackObject = stack;
+	continuation->stackOverflowMark = continuation->stackOverflowMark2 = J9JAVASTACK_STACKOVERFLOWMARK(stack);
+
+	frame = ((J9SFJNINativeMethodFrame*)stack->end) - 1;
+	frame->method = NULL;
+	frame->specialFrameFlags = 0;
+	frame->savedCP = NULL;
+	frame->savedPC = (U_8*)(UDATA)J9SF_FRAME_TYPE_END_OF_STACK;
+	frame->savedA0 = (UDATA*)(UDATA)J9SF_A0_INVISIBLE_TAG;
+	continuation->sp = (UDATA*)frame;
+	continuation->literals = (J9Method*)0;
+	continuation->pc = (U_8*)J9SF_FRAME_TYPE_JNI_NATIVE_METHOD;
+	continuation->arg0EA = (UDATA*)&frame->savedA0;
+	continuation->stackObject->isVirtual = TRUE;
+
+	J9VMJDKINTERNALVMCONTINUATION_SET_VMREF(currentThread, continuationObject, continuation);
+
+	/* GC Hook to register Continuation object */
+end:
+	return result;
+}
+
+void JNICALL
+resumeContinuation(J9VMThread *currentThread, J9VMContinuation *continuation)
+{
+	J9VMEntryLocalStorage newELS;
+	newELS = continuation->entryLocalStorage;
+	newELS.oldEntryLocalStorage = currentThread->entryLocalStorage;;
+	currentThread->entryLocalStorage = &newELS;
+
+	if (NULL != newELS.oldEntryLocalStorage) {
+		/* Assuming oldELS > newELS, bytes used is (oldELS - newELS) */
+		UDATA freeBytes = currentThread->currentOSStackFree;
+		UDATA usedBytes = ((UDATA)newELS.oldEntryLocalStorage - (UDATA)&newELS);
+		freeBytes -= usedBytes;
+		currentThread->currentOSStackFree = freeBytes;
+
+		if ((IDATA)freeBytes < J9_OS_STACK_GUARD) {
+			if (J9_ARE_NO_BITS_SET(currentThread->privateFlags, J9_PRIVATE_FLAGS_CONSTRUCTING_EXCEPTION)) {
+				setCurrentExceptionNLS(currentThread, J9VMCONSTANTPOOL_JAVALANGSTACKOVERFLOWERROR, J9NLS_VM_OS_STACK_OVERFLOW);
+				currentThread->currentOSStackFree += usedBytes;
+			}
+		}
+	}
+
+	VM_OutOfLineINL_Helpers::restoreInternalNativeStackFrame(currentThread);
+	VM_OutOfLineINL_Helpers::returnSingle(currentThread, JNI_TRUE, 0);
+
+	currentThread->returnValue = J9_BCLOOP_EXECUTE_BYTECODE;
+
+	c_cInterpreter(currentThread);
+
+	restoreCallInFrame(currentThread);
+}
+
+BOOLEAN
+enterContinuation(J9VMThread *currentThread, j9object_t continuationObject)
+{
+	BOOLEAN result = TRUE;
+	jboolean started = J9VMJDKINTERNALVMCONTINUATION_STARTED(currentThread, continuationObject);
+	J9VMContinuation *continuation = J9VMJDKINTERNALVMCONTINUATION_VMREF(currentThread, continuationObject);
+
+	Assert_VM_Null(currentThread->currentContinuation);
+
+	VM_ContinuationHelpers::swapFieldsWithContinuation(currentThread, continuation);
+	currentThread->currentContinuation = continuation;
+
+	Assert_VM_notNull(currentThread->currentContinuation);
```

Comment:
Is this assertion necessary, or just need a read/write barrier instead?

Reply:
similarl a notnull asertion on the continuation before swap fields would be helpful